### PR TITLE
Viewpoint adaptation bug on generics?

### PR DESCRIPTION
**Pull request by AndrewShf**
See if any testcases fail...
TODO: explain the problem and the approach.

### CODE REVIEW

File: framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java [hunk 1]
```diff
@@ -64,9 +64,7 @@ public void viewpointAdaptMember(
         if (!shouldAdaptMember(memberType, memberElement)) {
             return;
         }
-
-        AnnotatedTypeMirror decltype = atypeFactory.getAnnotatedType(memberElement);
-        AnnotatedTypeMirror combinedType = combineTypeWithType(receiverType, decltype);
+        AnnotatedTypeMirror combinedType = combineTypeWithType(receiverType, memberType);
```
Comment (AndrewShf) [inline framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java:67]: replacing `memberType` with `decltype` isn't nice, as `membertype` is a concrete type and already be performed with substitution. `combinetypewithtype` should be the one to perform substition.

File: framework/tests/viewpointtest/AnnoOnTypeVariableUse.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import viewpointtest.quals.*;
```
Comment (AndrewShf) [inline framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:1]: Treatment to type variables?
RDM: do receiver-dependent viewpoint adaptation.
add a new qualifier LUB: to do LUB of the type argument and the class, have transitive immutability
RI (mutable/immutable): no rdm, no lub effects.

Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:1]: think about the ri subtitution.

File: framework/tests/viewpointtest/AnnoOnTypeVariableUse.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+@SuppressWarnings("cast.unsafe.constructor.invocation")
+class Demo<E> {
+    @ReceiverDependentQual E element;
```
Comment (AndrewShf) [inline framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:5]: think about rdm use on typevar!

File: framework/tests/viewpointtest/AnnoOnTypeVariableUse.java (hunk 2 above)
Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:1]: RI uses type argument information / mutable ignores type argument.

Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:1]: Box < T> {
    @mutable T f;
}
@Immutable
class C {
}
Box < @Immutable C>  a;
a.f --> mutable which contradicts to the definition of C.
well-form rules check / or change the semantics on the type variable use.

File: framework/tests/viewpointtest/AnnoOnTypeVariableUse.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import viewpointtest.quals.*;
+
```
Comment (AndrewShf) [inline framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:2]: So who has the priority? substitution or viewpoint adaptation?
substitution replaces the type params with the concrete type arguments.
viewpoint adaptation adapts the declared type to the receiver type.

Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:2]: Screen Shot 2023-05-08 at 11 04 29 AM
there's no viewpoint adaptation for rdq generics at all?

this code has a similar error 
```
@SuppressWarnings("cast.unsafe.constructor.invocation")
class Demo<E> {
    @A E element;

    static void test() {
        Demo<@B Element> d = new @A Demo<@B Element>();
        // d.element = @A |> @RDQ = @A
        // thus expects no error here
        d.element = new @A Element();
    }
}

class Element {
    int f = 1;
}
```

Maybe the `abstractViewpointAdapter` was designed for GUT, and didn't consider annotations on typevars?

Comment (wmdietl) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:2]: @AndrewShf Yes, this is definitely colored by how GUT formalized viewpoint adaptation. Annotations on type variable uses was not a part of GUT, so there is no clear semantics. As discussed, please think about what a good semantics would be and how we could implement it.

Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:2]: I have a design for PICO, as sent via the email. But I don't quite understand for a base type system, what kind of features we want in viewpoint adaptation generics. If allowing qualifiers on type arguments, type parameters, and field declarations, there could be multiple interpretations for a typevar use. Which one should we choose, for the general type system? I find it unclear.

Comment (AndrewShf) [reply framework/tests/viewpointtest/AnnoOnTypeVariableUse.java:2]: LUB?

**Comment by AndrewShf** [comment]
reopened in another pr.

**Pull request closed without merging**